File: MyGame/app/src/main/java/com/example/mygame/domain/visitor/ScreenCollisionVisitor.kt
package com.example.mygame.domain.visitor

import com.example.mygame.domain.Enemy
import com.example.mygame.domain.IVisitor
import com.example.mygame.domain.Platform
import com.example.mygame.domain.Screen
import com.example.mygame.domain.bonus.Jetpack
import com.example.mygame.domain.bonus.Shield
import com.example.mygame.domain.bonus.Spring
import com.example.mygame.domain.bullet.Bullet
import com.example.mygame.domain.platform.MovingPlatformOnX
import com.example.mygame.domain.player.Player

class ScreenCollisionVisitor(private val screen: Screen) : IVisitor {
    override fun visit(player: Player) {
        if (isCollidePlayerWithScreen(player)) {
            player.movingThroughScreen(screen)
        }
    }

    override fun visit(platform: Platform) {
        if (platform is MovingPlatformOnX && isCollidePlatformWithBoundsOfScreen(platform)) {
            platform.changeDirectionX(screen)
        }

        if (isOutPlatformBelowBottomOfScreen(platform)) {
            platform.isDisappeared = true
        }
    }

    override fun visit(spring: Spring) {}
    override fun visit(shield: Shield) {}
    override fun visit(jetpack: Jetpack) {}
    override fun visit(enemy: Enemy) {}

    override fun visit(bullet: Bullet) {
        if (screen.top > bullet.bottom) {
            bullet.isDisappeared = true
        }
    }

    private fun isCollidePlayerWithScreen(player: Player) : Boolean {
        return player.right < screen.left || player.left > screen.right
    }

    private fun isCollidePlatformWithBoundsOfScreen(platform: Platform) : Boolean {
        return platform.right >= screen.right || platform.left <= screen.left
    }

    private fun isOutPlatformBelowBottomOfScreen(platform: Platform) : Boolean {
        return platform.top > screen.bottom
    }
}

File: GameServer/src/Collision/Visitor/ScreenCollisionVisitor.php
<?php

namespace App\Collision\Visitor;

use App\Model\Enemy;
use App\Model\Bullet;
use App\Model\Player;
use App\Model\Screen;
use App\Model\Platform;
use App\Model\Bonus\Shield;
use App\Model\Bonus\Spring;
use App\Collision\IVisitor;
use App\Model\EndZone;
use App\Model\Platform\MovingPlatformOnX;

class ScreenCollisionVisitor implements IVisitor
{
    private $screen;

    public function __construct(Screen $screen)
    {
        $this->screen = $screen;
    }

    public function visitPlayer(Player $player): void
    {
        if ($this->isCollidePlayerWithScreen($player)) {
            $player->movingThroughScreen($this->screen);
        }
    }

    public function visitPlatform(Platform $platform): void
    {
        if ($this->isOutPlatformBelowBottomOfScreen($platform)) {
            $platform->setIsDisappeared(true);
        }
    }

    public function visitSpring(Spring $spring): void
    {
    }

    public function visitEndZone(EndZone $endZone): void {
    }

    private function isCollidePlayerWithScreen(Player $player): bool
    {
        return $player->getX() < $this->screen->left || $player->getX() > $this->screen->right;
    }

    private function isCollidePlatformWithBoundsOfScreen(Platform $platform): bool
    {
        return $platform->getRight() >= $this->screen->right || $platform->getLeft() <= $this->screen->left;
    }

    private function isOutPlatformBelowBottomOfScreen(Platform $platform): bool
    {
        return $platform->getTop() > $this->screen->bottom;
    }
}
